dbms/src/Interpreters/Context.cpp

```diff
@@ -245,15 +245,12 @@ struct ContextShared
             return;
         shutdown_called = true;
 
-        {
-            std::lock_guard lock(mutex);
 
-            /** After this point, system logs will shutdown their threads and no longer write any data.
-            * It will prevent recreation of system tables at shutdown.
-            * Note that part changes at shutdown won't be logged to part log.
-            */
-            system_logs.reset();
-        }
+        /** At this point, system logs will flush accumulated data, then shutdown their threads and no longer write any data.
+        * It will prevent recreation of system tables at shutdown.
+        * Note that part changes at shutdown won't be logged to part log.
+        */
+        system_logs.reset();
```

Comment:
I think locking mutex here is unnecessary, because anyway we will wait here until system logs will join their threads in shutdown method and there is no concurrency on system_logs. 
But this lock prevents to flush logs on shutdown.

Reply:
Unfortunately the data race is still here:
`MergeTreeData::removePartsFinally() --> DB::Context::getPartLog()`
vs
`DB::ContextShared::shutdown() --> system_logs.reset()`
So release must be done under lock.

But it is not really necessary to destroy the system_logs object here, we only need to shut it down: `system_logs->shutdown();`

I suggest we add this comment as well:
`After system_logs have been shut down it is guaranteed that no system log table gets created or written to.`

Reply:
It is also safe to leave destroying of the `system_logs` object to `DB::ContextShared::~ContextShared()`